src/mscorlib/src/System/Collections/Generic/List.cs

```diff
@@ -247,7 +246,7 @@ public void Add(T item)
             var array = _items;
             var size = _size;
             _version++;
-            if ((uint)size < (uint)array.Length)
+            if (array.IndexInRange(size))
```

Comment:
> If signed test was performed, the range check just has to test its >= 0, rather than comparing to length.

Yes, I was thinking about that recently. I'm not sure if it's useful, see below.

> The examples I was looking at the generally always ended up as a memory access; so probably dominated by L1 cache hit? (as already checked). Maybe if it kept the length in register from earlier lookup and test that would be most of the cost gone.

Usually the CSE picks up the duplicate length load so you end up with a single memory access. Something like:
```asm
mov eax, dword ptr [rdx+8]
cmp ecx, eax
jge (throw ArgumentException)
cmp ecx, eax
jae (range_check_failed)
```
What do we gain by eliminating the range check? 

We get rid of a `cmp` and a `jcc` but these are cheap instructions. The `jcc` should be correctly predicted and they may end up being executed in parallel with other instructions, they're not part of a dependency chain.

If there are no other range checks in the method we also get rid of a call to the range check failed helper. Good for overall program size but 0 impact on performance.

We may free up a register because now we can put the memory load in the `cmp` like you show in your example. This is where changing the range check to `i >= 0` could be useful because we no longer have to keep the array length in a register (or reload it from memory). Nice but usually insignificant.